pywinusb/hid/core.py

```diff
@@ -906,12 +924,16 @@ def __init__(self, hid_object, raw_report_size):
 
         def abort(self):
             """Stop collectiong reports."""
-            if not self.__active:
-                return
-            self.__abort = True
-            if self.is_alive() and self.__h_read_event:
-                # force overlapped events competition
-                winapi.SetEvent(self.__h_read_event)
+            with self.__abort_lock:
+                if not self.__abort and self.__h_read_event:
+                    # force overlapped events competition
+
+                    # The abort variable must be set to true
+                    # before sending the event, otherwise
+                    # the reader thread might skip
+                    # CancelIo
+                    self.__abort = True
+                    winapi.SetEvent(self.__h_read_event)

```

Comment:
An alternative to sending an event here is to call CancelIoEx.  If it is used then the read event can be local to the run function.  It would also make the code a little more correct, since even after an IO operation is canceled, it can still have a side effect and write to the overlapped structure it was passed.  To make sure the operation has been canceled, the code needs to wait until the event is signaled by ReadFile.  If the operation was canceled then the result should be ERROR_OPERATION_ABORTED.

One downside with this is that CancelIoEx on exists on Windows Vista and above, so this would not run on XP machines.  Let me know what you think.  I could also make the modification in another PR in the future.  As the code is right now, it resolves all the issues I was seeing.